subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,220 @@
+#define CB_BUF_SIZE	CONFIG_IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+
+static const uint8_t magic[] = {0x45, 0x6d, 0x31, 0x6c, 0x31, 0x4b,
+				0x30, 0x72, 0x6e, 0x33, 0x6c, 0x69, 0x34};
+BUILD_ASSERT(sizeof(magic) <= CB_BUF_SIZE);
+BUILD_ASSERT(CB_BUF_SIZE <= UINT8_MAX);
+
+struct backend_data_t {
+	/* Tx/Rx buffers. */
+	struct icmsg_buf *tx_ib;
+	struct icmsg_buf *rx_ib;
+
+	/* Backend ops for an endpoint. */
+	const struct ipc_service_cb *ops;
+
+	/* General */
+	struct k_work mbox_work;
+	atomic_t state;
+};
+
+struct backend_config_t {
+	uintptr_t tx_shm_addr;
+	uintptr_t rx_shm_addr;
+	size_t tx_shm_size;
+	size_t rx_shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	uint8_t cb_buffer[CB_BUF_SIZE] __aligned(4);
+
+	atomic_t state = atomic_get(&dev_data->state);
+	int len = icmsg_buf_read(dev_data->rx_ib, cb_buffer, CB_BUF_SIZE);
+
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
+		return;
+	} else if (len <= 0) {
+		return;
+	}
+
+	if (state == ICMSG_STATE_READY) {
+		dev_data->ops->received(cb_buffer, len, NULL);
+
+		/* Reading with NULL buffer to know if there are data in the buffer to be read. */
+		len = icmsg_buf_read(dev_data->rx_ib, NULL, 0);
+		if (len > 0) {
+			(void)k_work_submit(&dev_data->mbox_work);
+		}
+	} else {
+		__ASSERT_NO_MSG(state == ICMSG_STATE_BUSY);
+		if (len != sizeof(magic) || memcmp(magic, cb_buffer, len)) {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		dev_data->ops->bound(NULL);
+		atomic_set(&dev_data->state, ICMSG_STATE_READY);
+	}
+}
+
+static void mbox_callback(const struct device *instance, uint32_t channel,
+			  void *user_data, struct mbox_msg *msg_data)
+{
+	struct backend_data_t *dev_data = user_data;
+
+	(void)k_work_submit(&dev_data->mbox_work);
+}
+
+static int mbox_init(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	int err;
+
+	k_work_init(&dev_data->mbox_work, mbox_callback_process);
+
+	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, dev_data);
+	if (err != 0) {
+		return err;
+	}
+
+	return mbox_set_enabled(&conf->mbox_rx, 1);
+}
+
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	struct ipc_ept *ep = (struct ipc_ept *)token;
+	int ret;
+
+	if (!atomic_cas(&dev_data->state, ICMSG_STATE_OFF, ICMSG_STATE_BUSY)) {
+		/* Already registered. This backend supports single ep. */
+		return -EALREADY;
+	}
+
+	ep->instance = instance;
+	dev_data->ops = &cfg->cb;
+
+	if (!dev_data->ops->received || !dev_data->ops->bound) {
+		__ASSERT_NO_MSG(false);
+		return -EINVAL;
```

Comment:
Is this really invalid? I can't find such restriction in API description.

Reply:
it would be rather pointless IPC if received is not there... 
we could question bound - but how one can know when sending is safe?

I believe it's better to not accept missing callbacks.

Reply:
Actually @pdunaj i can see one case when the callbacks are not essential, its when one side of the communication channel does not expect to receive any data and is only focused on sending. Bound is kinda optional so i could also make it work. I will adapt it so it could work without callbacks.

Reply:
Also I can see that the reference implementation (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c) gracefully handles when callbacks are set to `NULL`. I think we should keep consistent behavior between backend implementations.